src/extlibs/hrvo/visualizer/hrvo_visualizer.py

```diff
@@ -0,0 +1,175 @@
+import matplotlib.pyplot as plt
+from matplotlib.animation import FuncAnimation, PillowWriter
+from matplotlib.patches import Circle
+import pandas as pd
+
+# Constants
+robot_radius = 0.09
+frame_rate = 30
+frame_length_ms = 1000.0 / frame_rate
+play_back_speed = 1.0
+div_a_field_width = 9.0
+div_a_field_length = 12.0
+default_robot_color = "royalblue"
+collided_robot_color = "red"
+
+
+def animate_robots(robot_pos_df, gif_output_file=None):
+    num_frames = robot_pos_df["frame"].max() + 1
+    num_robots = robot_pos_df["robot_id"].max() + 1
+    default_line_alpha = 0.3
+    is_paused = False
+
+    # Calculate plot size
+    max_robot_x_pos = robot_pos_df["x"].max()
+    min_robot_x_pos = robot_pos_df["x"].min()
+    max_robot_y_pos = robot_pos_df["y"].max()
+    min_robot_y_pos = robot_pos_df["y"].min()
+    plot_scale_constant = 0.125
+    x_offset = max(plot_scale_constant * (max_robot_x_pos - min_robot_x_pos), 1.0)
+    y_offset = max(plot_scale_constant * (max_robot_y_pos - min_robot_y_pos), 1.0)
+
+    def setup_plot():
+        """Initial drawing of the scatter plot."""
+        for robot in robot_list:
+            ax.add_patch(robot)
+
+        for line in line_list:
+            line.set_data([], [])
+
+        return robot_list + line_list + [time_text] + [robot_id_text]
+
+    def update(frame):
+        """Update plot for new frame"""
+        if frame >= num_frames:
+            return robot_list + line_list + [time_text] + [robot_id_text]
+
+        # Current frame, dataframe filter
+        curr_frame_df = robot_pos_df[robot_pos_df["frame"] == frame]
+
+        # Update robot positions
+        for robot_id in range(len(robot_list)):
+            robot = curr_frame_df[curr_frame_df["robot_id"] == robot_id]
+            robot_list[robot_id].center = (
+                float(robot["x"].head(1)),
+                float(robot["y"].head(1)),
+            )
+
+            # Update robot color if it collides with another robot
+            other_robot_id = int(robot["has_collided"])
+            if other_robot_id != -1:
+                print(f"went into for loop with other robot {other_robot_id}\n{robot}")
+                robot_list[robot_id].set_facecolor(collided_robot_color)
+
+                if other_robot_id > robot_id:
+                    other_robot = curr_frame_df[
+                        curr_frame_df["robot_id"] == other_robot_id
+                    ]
+                    relative_vel_x = float(other_robot["velocity_x"]) - float(
+                        robot["velocity_x"]
+                    )
+                    relative_vel_y = float(other_robot["velocity_y"]) - float(
+                        robot["velocity_y"]
+                    )
+                    relative_speed = pow(
+                        pow(relative_vel_x, 2) + pow(relative_vel_y, 2), 0.5
+                    )
```

Comment:
```
difference = robot.subtract(other_robot)
relative_speed = math.hypot(difference["velocity_x"], difference["velocity_y"])
```

Just a suggestion on how this can be a bit more concise

haven't tried it, so there is a possibility that it doesn't work

Reply:
Wasn't able to get `.subtract` working. Kept returning NAN. But updated to use `math.hypot(...)`